### PR TITLE
rebuilding golden resources with survivorship in partitions

### LINKED ISSUE #5956: Rebuilding golden resources (mdm) with survivorship rules in partitioned environments does not work
Reproduction Steps
* create 2 partitions (one for regular resources, one for golden)
* create at least 1 resource w/ golden resource link
* rebuild the golden resource with survivorship rules

### PR DESCRIPTION
**Pull request by TipzCM**
closes #5956

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/index/IdHelperService.java [hunk 1]
```diff
@@ -155,25 +155,26 @@ public IResourceLookup<JpaPid> resolveResourceIdentity(
 	public IResourceLookup<JpaPid> resolveResourceIdentity(
 			@Nonnull RequestPartitionId theRequestPartitionId,
 			String theResourceType,
-			String theResourceId,
+			final String theResourceId,
```
Comment (lukedegruchy) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/index/IdHelperService.java:158]: Ugh!  final should be the default in Java

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/index/IdHelperService.java [hunk 2]
```diff
@@ -155,25 +155,26 @@ public IResourceLookup<JpaPid> resolveResourceIdentity(
...
 			boolean theExcludeDeleted)
 			throws ResourceNotFoundException {
 		assert myDontCheckActiveTransactionForUnitTest || TransactionSynchronizationManager.isSynchronizationActive()
 				: "no transaction active";
 
-		if (theResourceId.contains("/")) {
-			theResourceId = theResourceId.substring(theResourceId.indexOf("/") + 1);
+		String resourceIdToUse = theResourceId;
+		if (resourceIdToUse.contains("/")) {
+			resourceIdToUse = theResourceId.substring(resourceIdToUse.indexOf("/") + 1);
```
Comment (lukedegruchy) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/index/IdHelperService.java:166]: Question:  I have a vague idea why you're maintaining a variable separate from the parameter but I'm curious what use case it solves

**Review by lukedegruchy — APPROVED** [review]

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/index/IdHelperService.java (hunk 1 above)
Comment (TipzCM) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/index/IdHelperService.java:158]: ironically, it is.... but if you change the variable, java will then say it's not final.

Which is why i'm not a huge fan of this new "minimalist" coding structure where the language devines what is desired. :(

At any rate, this is now fixed here

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/index/IdHelperService.java (hunk 2 above)
Comment (TipzCM) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/index/IdHelperService.java:166]: It doesn't solve any specific use case.

I just don't like functions that mutate variables.

It leads to problems later on that are hard to diagnose and can often be even harder to fix as layers upon layers of code start depending on this functionality.

An example i had fixed before was in our search code where we mutate the search parameters to do the count search first.... but this mutation is then relied upon for some parts of the actual search later on, but not other parts. Resulting in us kicking around 2 versions of the parameters and having to maintain both.

I'd rather this not get to that point.

**Review by tadgh — APPROVED** [review]

**Pull request merged**